crates/fuel-core/src/coins_query.rs

```diff
@@ -221,6 +274,168 @@ pub async fn random_improve(
     Ok(coins_per_asset)
 }
 
+pub async fn select_coins_to_spend(
+    CoinsToSpendIndexIter {
+        big_coins_iter,
+        dust_coins_iter,
+    }: CoinsToSpendIndexIter<'_>,
+    total: u64,
+    max: u16,
+    excluded_ids: &ExcludedCoinIds<'_>,
+    batch_size: usize,
+) -> Result<Vec<CoinsToSpendIndexEntry>, CoinsQueryError> {
+    const TOTAL_AMOUNT_ADJUSTMENT_FACTOR: u64 = 2;
+    if total == 0 && max == 0 {
+        return Ok(vec![]);
+    }
+
+    let adjusted_total = total.saturating_mul(TOTAL_AMOUNT_ADJUSTMENT_FACTOR);
+
+    let big_coins_stream = futures::stream::iter(big_coins_iter).yield_each(batch_size);
+    let dust_coins_stream = futures::stream::iter(dust_coins_iter).yield_each(batch_size);
+
+    let (selected_big_coins_total, selected_big_coins) =
+        big_coins(big_coins_stream, adjusted_total, max, excluded_ids).await?;
+
+    if selected_big_coins_total < total {
+        return Ok(vec![]);
+    }
+    let Some(last_selected_big_coin) = selected_big_coins.last() else {
+        // Should never happen.
+        return Ok(vec![]);
+    };
```

Comment:
t would be nice to return an error here=)

Reply:
Added a proper error and some more explanation in 87245d406d2aaefba627d9f9345b98ae68d2a696